### ISSUE TITLE
[PERF] Bad frame timings every few seconds on Radeon RX 5700 XT and Valve Index

### ISSUE THREAD

**Issue by lubosz** [ValveSoftware/SteamVR-for-Linux#245]
**Describe the bug**
Beat Saber runs well most of the time, but is unplayable on high levels due to the frequent drops.

While in the Beat Saber main menu `WaitGetPoses` usually takes ~4.5ms, but regularly drops to ~12.5 (as seen in the `beat-saber-main-menu-bad.trace`). In the `beat-saber-main-menu-onebigdrop.trace` the `WaitGetPoses` took 75ms once.
The problem seems to be hard to trace, since the frame timing improves when gpuviz is tracing.

**To Reproduce**
1. Start Beat Saber
2. Performance drops occur as early as the "Safety Warning Screen"

**Expected behavior**
Constant frame times of under 6.9ms.

**System Information (please complete the following information):**
- Distribution: Arch
- SteamVR version: 1.7.15
- Steam client version: Sep 10 2019
- Opted into Steam client beta?: Yes
- Graphics driver version: Mesa 19.3.0-devel (git-9466e4cfab) (LLVM 10.0.0)
- Kernel: `drm-next` at `9a60b2990d6c2b7ab935fe0a5cc274de67d98bed` with fsync patches applied.
- Gist for SteamVR System Information: [Gist](https://gist.github.com/lubosz/4ccc70f88697a8c90f42cb0b4cf2b236)

**Performance Data (optional, but very helpful):**
- CPU: CPU Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
  (Using `pti=off spectre_v2=off` kernel parameters)
- GPU: PowerColor Radeon RX 5700 XT "Red Devil"
- Gist for perf top: [Gist](https://gist.github.com/lubosz/e221902f19f35748c42135a1f40ca245)
- Link to gpuvis traces:
  [beat-saber-main-menu-bad.7z](https://drive.google.com/open?id=11c3w1UTK2lRjKV7MTv4324ALrUADP8Ab)
  [beat-saber-main-menu-onebigdrop.7z](https://drive.google.com/open?id=1X0G9Vn08gTocyJ9XHRnPvWatjMRZ2CU8)

**Screenshots**

### Beat Saber Valve Index

The small frame timings graph in SteamVR settings shows pink and orange frames frequently.

![beat-saber-brief-frame-timing](https://user-images.githubusercontent.com/295622/64921651-03471f80-d7c6-11e9-953b-f6a2a5cdb790.png)

![beat-saber-brief-frame-timing2](https://user-images.githubusercontent.com/295622/64921652-06421000-d7c6-11e9-9439-45b243a691ef.png)

In ~1200 frames you can clearly see the inhomogeneity of frame times in the advanced timings details view.
![beat-saber-timing-detail-all2](https://user-images.githubusercontent.com/295622/64921690-6d5fc480-d7c6-11e9-9be3-7a14e0ca0376.png)
The spikes can be tracked down in `WaitGetPoses` and `Application GPU (pre)`.
![beat-saber-timing-detail-filtered2](https://user-images.githubusercontent.com/295622/64921691-6fc21e80-d7c6-11e9-8beb-5509d4cd7d36.png)

### hellovr_vulkan Valve Index
In `hellovr_vulkan` the impact of `Total GPU` comparably low. `WaitGetPoses` and `Compositor Render Start` seem to be affected.

![hellovr-timing-filtered4](https://user-images.githubusercontent.com/295622/64921732-11497000-d7c7-11e9-9e20-b99f62468833.png)

### Beat Saber Vive Pro
On the Vive Pro the issue does not seem as impactful as on the Index.
![beat-saber-timing-detail-filtered-vive-pro](https://user-images.githubusercontent.com/295622/64921707-c0d21280-d7c6-11e9-8212-eac710f49054.png)

**Additional context**
In hellovr_vulkan and hellovr_opengl the problem also occurs, but only when gpuviz is not tracing. It is still visible in the details frame timing graph.

The SteamVR compositor itself runs constantly at 0.9ms, without any drops.

On the Vive Pro the frame time is constantly 3.9 without drastic drops. hellovr_vulkan does not have any noticeable drops on the Vive Pro.

Additional setup information:
* Valve Index with current firmware.
* `"enableLinuxVulkanAsync": false` in `steamvr.vrsettings`
* I also ran `echo high > /sys/class/drm/card0/device/power_dpm_force_performance_level`
* Proton 4.11-4, `fsync: up and running.`

**Note:** Commenters who are also experiencing this issue are encouraged to include the "System Information" and "Perf Data" sections in their replies.

**Comment by lubosz** [comment]
The situation seems to be better with only occasional frame skips on today's Linux master brach at `ad062195731bea1624ce7160e79e0fcdaa25c1b5` with fsync patches applied.
```
Linux bcube 5.3.0-gdcb8cfbd8fe9 #5 SMP PREEMPT Tue Sep 17 13:09:17 CEST 2019 x86_64 GNU/Linux
```

**Comment by lostgoat** [comment]
Hey @lubosz ,

It seems like the gpuvis captures don't fully support Navi yet so the data is incomplete. But from what I can piece together from the SteamVR side of the trace I think this is something that might be fixed in 1.8.2

Can you give it a try and let me know if the issue is still present?

**Comment by lubosz** [comment]
Hi @lostgoat ,

thanks for the reply. 
The performance improved with updating mesa to commit `05da025f350cb83e1125d6a931ccf9a87223bb69`, which added a bunch of radv patches by Samuel Pitoiset.
Updating Proton 4.11-5 also subjectively improved the performance further.
Tested on SteamVR 1.8.2, where I reset to default settings, so with Async Reprojection enabled again.

I didn't measure the impact of each update.

The performance is still not perfect, a bit worse than on Windows with occasional drops, but the game is playable and I was able to finish a game at Expert+.

For now I will close this and wait until the Navi stack matures a bit, and will resubmit an issue if the performance is unsatisfactory with release versions.

To recap, these versions work fine for me, all master branches.
`linux` @ `ad062195731bea1624ce7160e79e0fcdaa25c1b5` with fsync patches applied
`mesa` @ `05da025f350cb83e1125d6a931ccf9a87223bb69`
`SteamVR`  @ Beta `1.8.2`
`Proton` @ `4.11-5`

Update:
On SteamVR 1.8.3 the compositor only (without SteamVR Home) runs at 0.9ms with Beat Saber at ~3.6ms,
![Screenshot from 2019-09-19 10-12-48](https://user-images.githubusercontent.com/295622/65225868-2a536900-dac6-11e9-9f8d-6f78160d9261.png)
